Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set

anjali411 opened the pull request:
Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61925 Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set**

bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64), complex_float_tensor.conj().imag.view(torch.int32) now doesn't return a view but instead errors out

Differential Revision: [D29819288](https://our.internmc.facebook.com/intern/diff/D29819288)

anjali411 committed:
Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set

[ghstack-poisoned]

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"




[ghstack-poisoned]

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"

Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946


bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64) now doesn't return a view but instead errors out

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -144,6 +144,10 @@ Tensor view_dtype(const Tensor& self, ScalarType dtype) {
   TORCH_CHECK(self.element_size() == static_cast<int64_t>(type_meta.itemsize()),
     "Viewing a tensor as a new dtype with a different number of bytes per element is not supported.");
+  TORCH_CHECK(!self.is_conj(),
+    "torch.Tensor.view is not supported for conjugate view tensors when converting to non-complex dtype.");
```
You're not checking that this is converting to a non-complex dtype.
Is it ok under the assumption that there is only one complex dtype of each size?

anjali411 replied:
there are two possible cases:
1. complex_dtype1 -> complex_dtype1 (in which cases it returns from the first if)
2. complex_dtype1 -> complex_dtype2 (which would error out from the TORCH_CHECK above)

anjali411 replied:
offline sync: Disallow conversion of a conjugate view tensor to a different dtype since we are interpreting memory to a different format and it doesn't make sense to do that when the true value is not saved in memory as is the case for conjugate view tensors.

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -124,6 +124,16 @@ PyObject* tensor_to_numpy(const at::Tensor& tensor) {
         "Use tensor.detach().numpy() instead.");
   }
+  if (tensor.is_conj()) {
+    throw std::runtime_error(
```
We should never throw plain runtime_error, you should use `TORCH_CHECK()`

albanD commented on test/test_view_ops.py:
```diff
@@ -1493,6 +1493,12 @@ def test_view_all_dtypes_and_devices(self, device):
+        self.assertRaisesRegex(RuntimeError, "has conjugate bit set", lambda: torch.tensor([1+2j]).conj().numpy())
+        self.assertRaisesRegex(RuntimeError, "has negative bit set", lambda: torch.tensor([1+2j]).conj().imag.numpy())
+        self.assertRaisesRegex(RuntimeError, "not supported for conjugate view tensors", lambda: torch.tensor([1+2j]).conj().view(torch.float64))
+        self.assertRaisesRegex(RuntimeError, "not supported for tensors with negative bit set", lambda: torch.tensor([1+2j]).conj().imag.view(torch.int32))
```
To avoid the lines are too long problem, you can use these as context manager:
```python
with self.assertRaisesRegex(RuntimeError, "not supported for tensors with negative bit set"):
    torch.tensor([1 + 2j]).conj().imag.view(torch.int32)
```

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"

Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946


bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64), complex_float_tensor.conj().imag.view(torch.int32) now doesn't return a view but instead errors out

[ghstack-poisoned]

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -106,24 +106,28 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
-  }
-  if (tensor.device().type() != DeviceType::CPU) {
-    throw TypeError(
+  TORCH_CHECK(is_numpy_available(), "Numpy is not available");
```
Thanks for fixing the other occurences! :)

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -106,24 +106,28 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
-        "Use tensor.detach().numpy() instead.");
-  }
+
+  TORCH_CHECK(tensor.layout() == Layout::Strided,
```
Can you use `TORCH_CHECK_TYPE()` here so that the error type remains the same?
Same for the device above.

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819288).

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"

Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946


bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64), complex_float_tensor.conj().imag.view(torch.int32) now doesn't return a view but instead errors out

Differential Revision: [D29819288](https://our.internmc.facebook.com/intern/diff/D29819288)

[ghstack-poisoned]

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -106,24 +106,30 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
-  }
+  TORCH_CHECK(is_numpy_available(), "Numpy is not available");
+
+  TORCH_CHECK(tensor.device().type() == DeviceType::CPU,
```
This one was also a TypeError.

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"

Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946


bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64), complex_float_tensor.conj().imag.view(torch.int32) now doesn't return a view but instead errors out

Differential Revision: [D29819288](https://our.internmc.facebook.com/intern/diff/D29819288)

[ghstack-poisoned]

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -106,24 +106,30 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
+      " layout tensor to numpy.",
+      "convert the tensor to a strided layout first.");
+
+  TORCH_CHECK_TYPE(!(at::GradMode::is_enabled() && tensor.requires_grad()),
```
Wrong one :p

anjali411 committed:
Update on "Throw RuntimeError when numpy() is called on a tensor with conjugate or negative bit set"

Resolves https://github.com/pytorch/pytorch/issues/59945 and https://github.com/pytorch/pytorch/issues/59946


bc breaking note: Unlike before, complex_tensor.conj().numpy(),  complex_float_tensor.conj().view(torch.float64), complex_float_tensor.conj().imag.view(torch.int32) now doesn't return a view but instead errors out

Differential Revision: [D29819288](https://our.internmc.facebook.com/intern/diff/D29819288)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29819288).
